### ISSUE TITLE
awaiting a Promise recasts top variable to initial type

### ISSUE THREAD

**Issue by bondz** [Microsoft/TypeScript#19918]
**TypeScript Version:**  2.6.1, 2.7.0-dev.20171110

```ts
function test(a: boolean): false | Date {
    if (a) {
        return new Date();
    }

    return false;
}


async function test2() {
    let b = test(true);

    if (!b) {
        b = new Date();
    }
    
    await Promise.all([]).then(() => {
        b.getDate(); //Should compile.
            
    })
}
```

**Expected behavior:**
Should compile
**Actual behavior:**
Throws with `Property 'getDate' does not exist on type 'true'.`

---

This can be reproduced with any return type and even when type inference is used.

**Comment by aluanhaddad** [comment]
Likely duplicate of #13180. Also, see https://github.com/Microsoft/TypeScript/issues/9998 for detailed commentary and discussion.

Simple fix:
```ts
async function test2() {
  const b = test(true) || new Date();
  await Promise.all([]).then(() => {
      b.getDate(); // compiles.
  });
}
```

**Comment by bondz** [comment]
Thanls @aluanhaddad The simple fix works and looks like #9998 should cover this. Closing this ticket.
